pandas/core/frame.py

```diff
@@ -8869,11 +8869,30 @@ def update(
         if not isinstance(other, DataFrame):
             other = DataFrame(other)
 
-        other = other.reindex(self.index)
+        indexes_intersection = other.index.intersection(
+            self.index
+        )  # order is important
+        if not len(indexes_intersection):
+            raise ValueError(
+                "Can't update dataframe when other has no index in common with "
+                "this dataframe."
+            )
```

Comment:
Isn't this not backwards compatible? I think this case does not raise on main (in general), is that right?